#                                         Melbourne Housing Market

Get Dataset below : 
https://www.kaggle.com/anthonypino/melbourne-housing-market?select=Melbourne_housing_FULL.csv

# Overview :
            Melbourne is currently experiencing a housing bubble (some experts say it may burst soon). Maybe someone can 
            find a trend or give a prediction? Which suburbs are the best to buy in? Which ones are value for money? 
            Where's the expensive side of town? And more importantly where should I buy a 2 bedroom unit?
      Here we will look more into missing values, wayes to treat missing values & how much we have lost data after 
      computing missing values

# Missing values

In [12]:
import  pandas as pd
df = pd.read_csv("melbourne.csv")

In [13]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [14]:
df.shape

(23547, 21)

In [15]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

### Identifying Missing Values

In [17]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [18]:
round(100*(df.isnull().sum()/len(df.index)),3)

Suburb            0.000
Address           0.000
Rooms             0.000
Type              0.000
Price            21.875
Method            0.000
SellerG           0.000
Date              0.000
Distance          0.004
Postcode          0.004
Bedroom2         19.030
Bathroom         19.043
Car              19.646
Landsize         26.063
BuildingArea     57.455
YearBuilt        50.992
CouncilArea      33.512
Lattitude        18.278
Longtitude       18.278
Regionname        0.004
Propertycount     0.004
dtype: float64

In [19]:
# removing the three columns
df = df.drop('BuildingArea', axis=1)
df = df.drop('YearBuilt', axis=1)
df = df.drop('CouncilArea', axis=1)

In [20]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0  03-09-2016       2.5    3067.0       2.0       1.0  1.0     126.0   
1  03-12-2016       2.5    3067.0       2.0       1.0  1.0     202.0   
2  04-02-2016       2.5    3067.0       2.0       1.0  0.0     156.0   
3  04-02-2016       2.5    3067.0       3.0       2.0  1.0       0.0   
4  04-03-2017       2.5    3067.0       3.0       2.0  0.0     134.0   

   Lattitude  Longtitude             Regionname  Propertycount  
0   -37.8014    144.9958  Northern Metropolitan         4019.0  
1   -37.7996    144.9984  Northern Metropolitan         4019.0  
2   -37.8079    144.9934  Northern Metropolitan         4019.0  
3   -37.8114    145.0116  Northern Metropolitan         4019.0  
4   -37.8093    144.9944  Northern Metropolitan         4019.0

In [21]:
df.shape

(23547, 18)

In [22]:
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

### Treating Missing Values in Rows

In [25]:
df[df.isnull().sum(axis=1) > 5]

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St      3    h  1000000.0      S   
16            Abbotsford      18a Mollison St      2    t   745000.0      S   
19            Abbotsford  403/609 Victoria St      2    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22            Abbotsford    106/119 Turner St      1    u   481000.0     SP   
...                  ...                  ...    ...  ...        ...    ...   
23495  Templestowe Lower  1/207 Manningham Rd      2    u   550000.0     PI   
23499          Thornbury      1/128 Dundas St      2    t   770000.0     PI   
23502          Thornbury        111 Pender St      2    t   858000.0      S   
23508             Toorak   21/1059 Malvern Rd      2    u   720000.0     VB   
23531           Werribee         91 Latham St      4    h   540000.0     PI   

            SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19           Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21           Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22     Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
...             ...         ...       ...       ...       ...       ...  ...   
23495         Barry  26-08-2017      12.4    3107.0       NaN       NaN  NaN   
23499       McGrath  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23502        Jellis  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23508        Beller  26-08-2017       4.1    3142.0       NaN       NaN  NaN   
23531       Triwest  26-08-2017      14.7    3030.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
15          NaN        NaN         NaN  Northern Metropolitan         4019.0  
16          NaN        NaN         NaN  Northern Metropolitan         4019.0  
19          NaN        NaN         NaN  Northern Metropolitan         4019.0  
21          NaN        NaN         NaN  Northern Metropolitan         4019.0  
22          NaN        NaN         NaN  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23495       NaN        NaN         NaN   Eastern Metropolitan         5420.0  
23499       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23502       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23508       NaN        NaN         NaN  Southern Metropolitan         7217.0  
23531       NaN        NaN         NaN   Western Metropolitan        16166.0  

[4278 rows x 18 columns]

In [33]:
#### missing values more than 5 in each rows
len(df[df.isnull().sum(axis=1) > 5].index)

4278

In [39]:
### % wise missing values
round(100 *(len(df[df.isnull().sum(axis=1) > 5].index) / len(df.index)),3)

18.168

In [40]:
df = df[df.isnull().sum(axis=1) <= 5]

In [41]:
df.shape

(19269, 18)

In [47]:
round(100 * (df.isnull().sum() / len(df.index)),3)

Suburb            0.000
Address           0.000
Rooms             0.000
Type              0.000
Price            21.708
Method            0.000
SellerG           0.000
Date              0.000
Distance          0.000
Postcode          0.000
Bedroom2          1.054
Bathroom          1.069
Car               1.806
Landsize          9.648
Lattitude         0.135
Longtitude        0.135
Regionname        0.000
Propertycount     0.000
dtype: float64

In [49]:
df.shape

(19269, 18)

In [54]:
import numpy as np

# removing NaN Price rows
df = df[~np.isnan(df['Price'])]

In [55]:
df.shape

(15086, 18)

In [56]:
round(100 * (df.isnull().sum() / len(df.index)),3)

Suburb           0.000
Address          0.000
Rooms            0.000
Type             0.000
Price            0.000
Method           0.000
SellerG          0.000
Date             0.000
Distance         0.000
Postcode         0.000
Bedroom2         1.054
Bathroom         1.067
Car              1.763
Landsize         9.830
Lattitude        0.146
Longtitude       0.146
Regionname       0.000
Propertycount    0.000
dtype: float64

In [57]:
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

In [58]:
# removing NaNs in Landsize
df = df[~np.isnan(df['Landsize'])]

In [64]:
df.shape

(13603, 18)

In [65]:
round(100 * (df.isnull().sum() / len(df.index)),3)

Suburb           0.000
Address          0.000
Rooms            0.000
Type             0.000
Price            0.000
Method           0.000
SellerG          0.000
Date             0.000
Distance         0.000
Postcode         0.000
Bedroom2         0.000
Bathroom         0.007
Car              0.463
Landsize         0.000
Lattitude        0.162
Longtitude       0.162
Regionname       0.000
Propertycount    0.000
dtype: float64

In [66]:
# rows having Lattitude and Longitude missing
df[np.isnan(df['Lattitude'])]

Suburb                Address  Rooms Type      Price  \
2572              Burwood           23 Monica St      3    h   990000.0   
3257         Clifton Hill  3/268 Alexandra Pde E      1    u   363000.0   
4485            Footscray          483 Barkly St      3    t   781000.0   
5170         Hampton East           7 Seafoam St      4    t  1185000.0   
10745  Williamstown North         4/9 Adeline St      1    u   355000.0   
13223           Melbourne   1913/228 Abeckett St      3    u  1175000.0   
14008            Brooklyn          9 Richards Ct      3    h   750000.0   
14132     North Melbourne   13/201 Abbotsford St      2    t   755000.0   
14139      Oakleigh South            4 Druitt St      4    h  1205500.0   
14142      Oakleigh South        298 Warrigal Rd      3    h   799999.0   
14143      Oakleigh South            11 Yarra Ct      5    h  1200000.0   
14976            Essendon        9 Washington St      3    h  1520000.0   
15278              Seddon          60 Station Rd      4    h  1500000.0   
15561             Croydon       3 Silvergrass Ct      3    h   630000.0   
16358          Kensington     201/102 Rankins Rd      2    u   876000.0   
16564          Strathmore          48 Lebanon St      3    h   950000.0   
17616          Kensington     109/102 Rankins Rd      1    u   410000.0   
17620         Keysborough          19 Denmark Rd      5    h  1100000.0   
18141               Lalor           83 Rotino Cr      3    t   463000.0   
18614           Mickleham        17 Primavera Dr      4    h   610000.0   
18793             Wollert      13 Strathalbyn Ch      4    h   631000.0   
18995           Greenvale         40 Frontier Av      3    t   470000.0   

      Method        SellerG        Date  Distance  Postcode  Bedroom2  \
2572      VB      Fletchers  17-09-2016      11.7    3125.0       3.0   
3257       S  hockingstuart  27-06-2016       3.4    3068.0       1.0   
4485       S            Jas  27-11-2016       6.4    3011.0       3.0   
5170       S             RT  28-05-2016      14.5    3188.0       4.0   
10745      S        Sweeney  27-11-2016       8.9    3016.0       1.0   
13223     PI           Icon  29-04-2017       2.8    3000.0       3.0   
14008      S  hockingstuart  20-05-2017      10.9    3012.0       3.0   
14132     PI         Nelson  29-04-2017       2.3    3051.0       2.0   
14139      S       Woodards  22-04-2017      14.7    3167.0       4.0   
14142      S       Woodards  29-04-2017      14.7    3167.0       3.0   
14143     PI         Buxton  29-04-2017      14.7    3167.0       5.0   
14976      S           Brad  03-06-2017       7.5    3040.0       3.0   
15278     SP        Sweeney  03-06-2017       5.1    3011.0       4.0   
15561      S        McGrath  17-06-2017      23.0    3136.0       3.0   
16358      S        Rendina  24-06-2017       3.4    3031.0       2.0   
16564     PI      Considine  24-06-2017       8.2    3041.0       3.0   
17616     PI        Rendina  08-07-2017       3.4    3031.0       1.0   
17620     PI       VicHomes  08-07-2017      25.2    3173.0       5.0   
18141      S            HAR  12-08-2017      16.3    3075.0       3.0   
18614      S            Ray  15-07-2017      20.6    3064.0       4.0   
18793     PI  hockingstuart  15-07-2017      25.5    3750.0       4.0   
18995     SP          Barry  22-07-2017      20.4    3059.0       3.0   

       Bathroom  Car  Landsize  Lattitude  Longtitude  \
2572        2.0  2.0     263.0        NaN         NaN   
3257        1.0  1.0       0.0        NaN         NaN   
4485        2.0  2.0      98.0        NaN         NaN   
5170        3.0  1.0     300.0        NaN         NaN   
10745       1.0  1.0      46.0        NaN         NaN   
13223       3.0  2.0       0.0        NaN         NaN   
14008       1.0  2.0     667.0        NaN         NaN   
14132       1.0  1.0    1537.0        NaN         NaN   
14139       2.0  2.0     553.0        NaN         NaN   
14142       2.0  4.0     590.0        NaN  

In [68]:
df.loc[:,['Lattitude','Longtitude']].describe()
#df.loc[:, ['Lattitude', 'Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [69]:
round(100 * (df.isnull().sum() / len(df.index)),3)

Suburb           0.000
Address          0.000
Rooms            0.000
Type             0.000
Price            0.000
Method           0.000
SellerG          0.000
Date             0.000
Distance         0.000
Postcode         0.000
Bedroom2         0.000
Bathroom         0.007
Car              0.463
Landsize         0.000
Lattitude        0.162
Longtitude       0.162
Regionname       0.000
Propertycount    0.000
dtype: float64

In [72]:
df['Lattitude'].fillna((df['Lattitude'].mean()), inplace=True)
df['Longtitude'].fillna((df['Longtitude'].mean()), inplace=True)

In [73]:
round(100 * (df.isnull().sum() / len(df.index)),3)

Suburb           0.000
Address          0.000
Rooms            0.000
Type             0.000
Price            0.000
Method           0.000
SellerG          0.000
Date             0.000
Distance         0.000
Postcode         0.000
Bedroom2         0.000
Bathroom         0.007
Car              0.463
Landsize         0.000
Lattitude        0.000
Longtitude       0.000
Regionname       0.000
Propertycount    0.000
dtype: float64

In [74]:
df.shape

(13603, 18)

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13603 entries, 1 to 23545
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13603 non-null  object 
 1   Address        13603 non-null  object 
 2   Rooms          13603 non-null  int64  
 3   Type           13603 non-null  object 
 4   Price          13603 non-null  float64
 5   Method         13603 non-null  object 
 6   SellerG        13603 non-null  object 
 7   Date           13603 non-null  object 
 8   Distance       13603 non-null  float64
 9   Postcode       13603 non-null  float64
 10  Bedroom2       13603 non-null  float64
 11  Bathroom       13602 non-null  float64
 12  Car            13540 non-null  float64
 13  Landsize       13603 non-null  float64
 14  Lattitude      13603 non-null  float64
 15  Longtitude     13603 non-null  float64
 16  Regionname     13603 non-null  object 
 17  Propertycount  13603 non-null  float64
dtypes: flo

In [76]:
df.loc[:, ['Bathroom', 'Car']].describe()

Bathroom           Car
count  13602.000000  13540.000000
mean       1.534921      1.610414
std        0.691834      0.962244
min        0.000000      0.000000
25%        1.000000      1.000000
50%        1.000000      2.000000
75%        2.000000      2.000000
max        8.000000     10.000000

In [82]:
# converting to type 'category'
df['Car'] = df['Car'].astype('category')
df["Car"].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [84]:
# imputing NaNs by 2.0
df.loc[pd.isnull(df['Car']), ['Car']] = 2
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.00
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

In [85]:
df.shape

(13603, 18)

In [87]:
# converting to type 'category'
df['Bathroom'] = df['Bathroom'].astype('category')
df["Bathroom"].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [88]:
# imputing NaNs by 1
df.loc[pd.isnull(df['Bathroom']), ['Bathroom']] = 1
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64

In [89]:
df.shape

(13603, 18)

In [100]:
### Total lost data %
(len(df.index )/23547) * 100

57.769567248481756

# Observation : 
                            Out of 100 % [23547] dataset ,after missing values left 58 %[13603] of data.
                            And 42% [9944] data we have lost.
                            